### PR TITLE
wire storageversion

### PR DESCRIPTION
**Pull request by alexzielenski**
#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go [hunk 1]
```diff
@@ -263,6 +282,101 @@ func (s *DefaultStorageFactory) NewConfig(groupResource schema.GroupResource) (*
...
+	// Look up example in scheme to find all objects of the same Group-Kind
+	// Use the highest priority version for that group-kind whose lifecycle window
+	// includes the current emulation version.
+	// If no version is found, use the binary version
+	// (in this case the API should be disabled anyway)
+
+	//!TODO: Is there a better scheme to use here? Does this include
+	// aggregator scheme/apiextensions scheme? What about for aggregated apiservers?
+	// May have to augment constructor with list of schemes to try...
+	sch := legacyscheme.Scheme
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go]: Is the right thing to do here to add a scheme arg to `NewDefaultStorageFactory`? Do we need to check more than one scheme?

File: staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go [hunk 2]
```diff
@@ -263,6 +282,101 @@ func (s *DefaultStorageFactory) NewConfig(groupResource schema.GroupResource) (*
...
+	versions := sch.VersionsForGroupKind(schema.GroupKind{Group: gvks[0].Group, Kind: gvks[0].Kind})
+
+	//!TODO: what emulation version to use? Is it dependent upon component
+	// which depends upon source of the object?
+	// Are we mapping component version to a global k8s version?
+	//
+	// Purposefully using MinCompatibilityVersion rather than EmulationVersion
+	// for storage version determinations because we want the persisted resource
+	// to be compatible with MinCompatibilityVersion, not EmulationVersion.
+	compatibilityVersion := effectiveVersion.MinCompatibilityVersion()
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go]: We take effective version as an argument, how does the `MinCompatibilityVersion` in `EffectiveVersion` compare against the versions in the `APILifecycleIntroduced` methods? Is `APILifecycleIntroduced` always guaranteed to be relative to k8s binary versions (as opposed to the component it is from)

File: staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go [hunk 3]
```diff
@@ -263,6 +282,101 @@ func (s *DefaultStorageFactory) NewConfig(groupResource schema.GroupResource) (*
...
+			}
+		}
+
+		//!TODO: Throw error if resource doesn't conform to interface?
+		// (If user has compat mode set but we dont have lfiecycle data for the resource)
+		// NOTE: flowcontrol v1 has no lifecycle information, so it would always
+		// reach this point and use the binary version.
+		if !hasIntroduced && !hasRemoved {
+			continue
+		}
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go]: Had to add logic to ignore resources without lifecycle information, otherwise a v1 without any information is always at the front of the list and gets chosen. Should we enforce that all resources have lifecycle info?

File: pkg/controlplane/apiserver/config.go [hunk 4]
```diff
@@ -215,7 +218,7 @@ func BuildAuthorizer(ctx context.Context, s controlplaneapiserver.CompletedOptio
 // The peer endpoint leases are used to find network locations of apiservers for peer proxy
 func CreatePeerEndpointLeaseReconciler(c genericapiserver.Config, storageFactory serverstorage.StorageFactory) (reconcilers.PeerEndpointLeaseReconciler, error) {
 	ttl := controlplane.DefaultEndpointReconcilerTTL
-	config, err := storageFactory.NewConfig(api.Resource("apiServerPeerIPInfo"))
+	config, err := storageFactory.NewConfig(api.Resource("apiServerPeerIPInfo"), &corev1.Endpoints{})
```
Comment (alexzielenski) [inline pkg/controlplane/apiserver/config.go]: is this right

File: pkg/controlplane/instance.go [hunk 5]
```diff
@@ -283,7 +284,7 @@ func (c *Config) createLeaseReconciler() reconcilers.EndpointReconciler {
 	endpointsAdapter := reconcilers.NewEndpointsAdapter(endpointClient, endpointSliceClient)
 
 	ttl := c.ExtraConfig.MasterEndpointReconcileTTL
-	config, err := c.ExtraConfig.StorageFactory.NewConfig(api.Resource("apiServerIPInfo"))
+	config, err := c.ExtraConfig.StorageFactory.NewConfig(api.Resource("apiServerIPInfo"), &corev1.Endpoints{})
```
Comment (alexzielenski) [inline pkg/controlplane/instance.go]: is this right?

1. Should we subject this resource to compatibility versioning
2. Appeared to use corev1.Endpoints

File: pkg/registry/core/rest/storage_core.go [hunk 6]
```diff
@@ -321,7 +321,7 @@ func (p *legacyProvider) NewRESTStorage(apiResourceConfigSource serverstorage.AP
 func (c *Config) newServiceIPAllocators() (registries rangeRegistries, primaryClusterIPAllocator ipallocator.Interface, clusterIPAllocators map[api.IPFamily]ipallocator.Interface, nodePortAllocator *portallocator.PortAllocator, err error) {
 	clusterIPAllocators = map[api.IPFamily]ipallocator.Interface{}
 
-	serviceStorageConfig, err := c.StorageFactory.NewConfig(api.Resource("services"))
+	serviceStorageConfig, err := c.StorageFactory.NewConfig(api.Resource("services"), &corev1.Service{})
```
Comment (alexzielenski) [inline pkg/registry/core/rest/storage_core.go]: is this right (applying compatibility versioning to this special services storage)

**Review by alexzielenski — COMMENTED** [review]
Placed some comments where I am unclear on best path forward.

There is a lot of ripple to the many callsites of `GetRESTOptions` and `NewConfig`. The decision whether to supply `nil` or not depending upon whether this was a place we cared about the storage version (e.g. some calls just want resourceprefix or are for tests which should test against binary version)

File: staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go (hunk 1 above)
Comment (siyuanfoundation) [reply staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go]: should probably be consistent with `NewDefaultResourceEncodingConfig(scheme)`

File: staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go (hunk 2 above)
Comment (siyuanfoundation) [reply staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go]: This is similar to feature gate versions. Feature versions are always relative to the k8s binary versions. Even if the component has some different version, when calling k8s libraries, it should still have some map to translate that version to corresponding k8 version.

File: staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go [hunk 7]
```diff
@@ -263,6 +282,101 @@ func (s *DefaultStorageFactory) NewConfig(groupResource schema.GroupResource) (*
...
+			introducedVer := apimachineryversion.MajorMinor(uint(majorIntroduced), uint(minorIntroduced))
+			if introducedVer.GreaterThan(compatibilityVersion) {
+				continue
+			}
+		}
+
+		// If it was removed before current compatibility version, don't use it
+		removed, hasRemoved := exampleOfGVK.(removedInterface)
+		if hasRemoved {
+			majorRemoved, minorRemoved := removed.APILifecycleRemoved()
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go]: APILifecycleRemoved is when we stop *serving* a version, but we still understand how to decode / convert the version, so I don't think you need to exclude "removed" versions

File: staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go [hunk 8]
```diff
@@ -263,6 +282,101 @@ func (s *DefaultStorageFactory) NewConfig(groupResource schema.GroupResource) (*
...
+		//!TODO: Throw error if resource doesn't conform to interface?
+		// (If user has compat mode set but we dont have lfiecycle data for the resource)
+		// NOTE: flowcontrol v1 has no lifecycle information, so it would always
+		// reach this point and use the binary version.
+		if !hasIntroduced && !hasRemoved {
+			continue
+		}
+
+		// versions is returned in priority order, so just use first result
+		return gvk.GroupVersion(), nil
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go]: can drop this once we add lifecycle info to stable versions as well

File: staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go [hunk 9]
```diff
@@ -263,6 +282,101 @@ func (s *DefaultStorageFactory) NewConfig(groupResource schema.GroupResource) (*
...
+	}
+
+	return binaryVersionOfResource, nil
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go]: Getting here means we're serving a version that is unknown to the min-compatibility-version server.

This is only expected to happen when serving a brand new alpha API type (which doesn't happen by default), or when emulation-version and min-compatibility-version are several versions apart so a beta or GA API was being served which didn't exist at all in min-compatibility-version.

Would be good to explain this in a comment and test both of those scenarios

File: staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory_test.go [hunk 10]
```diff
@@ -244,3 +250,141 @@ func TestConfigs(t *testing.T) {
...
+	metav1.TypeMeta
+	metav1.ObjectMeta
+	Spec string
+}
+
+func (alphaAPIType) DeepCopyObject() runtime.Object {
+	return &alphaAPIType{}
+}
+
+func TestStorageFactoryCompatibilityVersion(t *testing.T) {
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory_test.go:271]: This test adds a lot of confidence for me that the override removals haven't broken anything.  Should we also include some non-corner cases for completeness?

File: staging/src/k8s.io/code-generator/cmd/prerelease-lifecycle-gen/prerelease-lifecycle-generators/status.go [hunk 11]
```diff
@@ -372,37 +377,53 @@ func (g *genPreleaseLifecycle) argsFromType(c *generator.Context, t *types.Type)
...
 		_, removedMajor, removedMinor, err = extractRemovedTag(t)
 		if err != nil {
 			return nil, err
 		}
 	}
-	a = a.
-		With("removedMajor", removedMajor).
-		With("removedMinor", removedMinor)
+
+	if !isGAVersion || hasRemoved {
```
Comment (jpbetz) [inline staging/src/k8s.io/code-generator/cmd/prerelease-lifecycle-gen/prerelease-lifecycle-generators/status.go:422]: Is there anywhere we can add test coverage for this isGAVersion condition? Is there sufficient coverage for the rest of these changes?

File: staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory_test.go (hunk 10 above)
Comment (alexzielenski) [reply staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory_test.go:271]: Added a few basic non-corner cases

File: staging/src/k8s.io/code-generator/cmd/prerelease-lifecycle-gen/prerelease-lifecycle-generators/status.go (hunk 11 above)
Comment (alexzielenski) [reply staging/src/k8s.io/code-generator/cmd/prerelease-lifecycle-gen/prerelease-lifecycle-generators/status.go:422]: There are no integration tests for this but I did add quite a few unit test cases for `argsFromType` with varying type arguments to show that how the emitted functions are controlled by the tags and package name

**Comment by alexzielenski** [comment]
added more tests, PTAL

**Comment by jpbetz** [comment]
/lgtm

Thanks @alexzielenski!
